From 755f10a7998e1b01c7777598eba6da63d86f446e Mon Sep 17 00:00:00 2001
From: William Douglas <william.douglas@intel.com>
Date: Wed, 5 Jul 2023 14:17:53 -0700
Subject: [PATCH] Check the state file pacdiscovery sets

If pacdiscovery did not run, then Pacrunner has no proxy script, thus proxy
resolution will not work - so just ignore any dbus connection and run libcurl.

Signed-off-by: Tudor Marcu <tudor.marcu@intel.com>
Signed-off-by: Arzhan Kinzhalin <arzhan.i.kinzhalin@intel.com>
Signed-off-by: Patrick McCarty <patrick.mccarty@intel.com>
---
 lib/url.c | 17 ++++++++++++-----
 1 file changed, 12 insertions(+), 5 deletions(-)

diff --git a/lib/url.c b/lib/url.c
index c888034..373c8b5 100644
--- a/lib/url.c
+++ b/lib/url.c
@@ -63,6 +63,8 @@
 /* Need dbus so we can query pacrunner for proxy */
 #include <dbus/dbus.h>
 #endif /* CURL_DISABLE_PROXY */
+#include <sys/stat.h>
+#define PROXY_STATE_FILE "/run/pacrunner/pac_active"
 
 #include <limits.h>
 
@@ -2187,6 +2189,7 @@ static char *detect_proxy(struct Curl_easy *data,
 #ifdef CURL_DISABLE_VERBOSE_STRINGS
   (void)data;
 #endif
+  struct stat sbuff;
 
   msnprintf(proxy_env, sizeof(proxy_env), "%s_proxy", conn->handler->scheme);
 
@@ -2224,13 +2227,17 @@ static char *detect_proxy(struct Curl_easy *data,
     }
     if(!proxy) {
 #endif
-      /* No protocol-specific proxy set in the environment.
-       *     Fallback to pacrunner autoproxy lookup */
-      autoproxy_ret = query_pacrunner_proxy(url,
+      /* Check that no proxy is set AND that pacdiscovery has set pacrunner
+       * with a PAC script */
+      if (stat(PROXY_STATE_FILE, &sbuff) == 0) {
+        /* No protocol-specific proxy set in the environment.
+         *     Fallback to pacrunner autoproxy lookup */
+        autoproxy_ret = query_pacrunner_proxy(url,
                           conn->host.name,
                           proxy_tmp);
-      if (!autoproxy_ret)
-          proxy = strndup(proxy_tmp, 256);
+        if (!autoproxy_ret)
+            proxy = strndup(proxy_tmp, 256);
+      }
     }
 
     if(!proxy) {
-- 
2.49.0

